### PR TITLE
Configurable admins, failover & decimal logic

### PR DESCRIPTION
**Pull request by alexroan**
- Decimal transformation if the underlying feed doesn't have 18 decimals
- Every price feed aggregator has a failover proxy
- Two levels of administrator:
    - `admin` - full control of all configuration functions
    - `failoverAdmin` - can't call configuration functions, but can failover
- `setAdmin` and `setFailoverAdmin` configuration functions added
- Using `require` statements rather than `ErrorReporter` handling

### CODE REVIEW

File: contracts/ChainlinkPriceOracle.sol [hunk 1]
```diff
@@ -37,28 +58,114 @@ contract ChainlinkPriceOracle is PriceOracle, OracleErrorReporter {
...
+     * @return Success code uint
+     */
+    function _setAdmin(address newAdmin) external returns (uint) {
+        // Check caller is admin
+        if (msg.sender != admin) {
+            return fail(Error.UNAUTHORIZED, FailureInfo.SET_ADMIN_OWNER_CHECK);
+        }
+
+        // Set new admin
+        admin = newAdmin;
```
Comment (krebernisak) [inline contracts/ChainlinkPriceOracle.sol:86]: Consider checking if the address actually changed, so the event is not raised in that case.

File: contracts/ChainlinkPriceOracle.sol [hunk 2]
```diff
@@ -37,28 +58,114 @@ contract ChainlinkPriceOracle is PriceOracle, OracleErrorReporter {
...
+     * @return Success code uint
+     */
+    function _setFailoverAdmin(address newFailoverAdmin) external returns (uint) {
+        // Check caller is admin
+        if (msg.sender != admin) {
+            return fail(Error.UNAUTHORIZED, FailureInfo.SET_FAILOVER_ADMIN_OWNER_CHECK);
+        }
+
+        // Set new admin
+        failoverAdmin = newFailoverAdmin;
```
Comment (krebernisak) [inline contracts/ChainlinkPriceOracle.sol:109]: Consider checking if the address actually changed, so the event is not raised in that case.

File: contracts/ChainlinkPriceOracle.sol [hunk 3]
```diff
@@ -28,37 +54,152 @@ contract ChainlinkPriceOracle is PriceOracle, OracleErrorReporter {
...
+
+        return uint(Error.NO_ERROR);
+    }
 
     /**
      * @notice Add a price feed for a cToken
+     * @dev Only callable by the administrator
      * @param cTokenAddress The address of the cToken
      * @param newPriceFeedAddress The address of the price feed
+     * @param failoverPriceFeedAddress The failover address
```
Comment (krebernisak) [inline contracts/ChainlinkPriceOracle.sol:122]: There are some params docs missing here, that would be great to have.

File: contracts/ChainlinkPriceOracle.sol [hunk 4]
```diff
@@ -28,37 +54,154 @@ contract ChainlinkPriceOracle is PriceOracle, OracleErrorReporter {
...
+        AggregatorDetails memory feedDetails = priceFeeds[address(cToken)];
+        require(address(feedDetails.feed) != address(0), "Price feed doesn't exist");
 
         // Get the price
-        int price = feed.latestAnswer();
+        int price = feedDetails.feed.latestAnswer();
         require(price >= 0, "Price cannot be negative");
-        return uint(price);
+
+        return uint(price).mul(uint(10)**feedDetails.decimals);
```
Comment (krebernisak) [inline contracts/ChainlinkPriceOracle.sol:63]: `uint(10)**feedDetails.decimals` could overflow as `uint8 decimals` can be a number like 255.

Consider limiting `decimals` to 32, for example.
That should be more than enough precision and will make this safe.

Comment (alexroan) [reply contracts/ChainlinkPriceOracle.sol:63]: ~Require max of 18 decimals as required by Compound?~

Do you think this is worth it, considering this configuration will be set and administered by governance, therefore highly scrutinized?

Comment (krebernisak) [reply contracts/ChainlinkPriceOracle.sol:63]: Does Compound require 18 decimals? I'm not sure on that one, but Chainlink aggregator contracts are configured with decimals ranging from 5 to 18 (max I think).

So we should probably be ok to limit it to 18 (please double-check the data), but we should also be aware of Compound requirements in cases where we have less than 18 decimals of precision.

Comment (kaleofduty) [reply contracts/ChainlinkPriceOracle.sol:63]: I'm in favor of adding a `require(decimals <= 18)` check in `_setPriceFeed`

**Pull request merged**